File: backend/src/models/report-model.ts
import { Expose, Transform, Type } from "class-transformer";
import { IsEnum } from "class-validator";
import { ObjectId } from "mongodb";
import { User } from "./user-model";
import { Mapper } from "../utils/mapper";
import { PostResponseDTO } from "./post-model";

export enum ReportStatus {
  ACTIVE = "ACTIVE",
  REJECTED = "REJECTED",
  ACCEPTED = "ACCEPTED",
  DELETED = "DELETED",
}

export class ReportModel {
  @Transform((value) => value.obj._id.toString())
  @Expose()
  _id?: ObjectId;

  @Expose()
  userId: string;

  @Expose()
  postId: string;

  @Expose()
  content?: string;

  @Expose()
  createdAt: Date;

  @Expose()
  @IsEnum(ReportStatus)
  status: ReportStatus;
}

export class ReportResponseDTO {
  @Expose()
  id: String;

  @Expose()
  userId: string;

  @Expose()
  postId: string;

  @Expose()
  content?: string;

  @Expose()
  createdAt: Date;

  @Expose()
  @IsEnum(ReportStatus)
  status: ReportStatus;

  @Expose()
  @Type(() => PostResponseDTO)
  post: PostResponseDTO;
}

export const mapToReportResponseDTO = (reportModel: ReportModel, postModel: PostResponseDTO): PostResponseDTO => {
  return Mapper.map(ReportResponseDTO, {
    ...reportModel,
    post: postModel,
    id: reportModel._id!.toString(),
  });
};

File: bilfind_app/lib/models/report_model.dart
import 'package:bilfind_app/models/post_model.dart';

class ReportModel {
  ReportModel(
      {required this.id,
      required this.userId,
      required this.postId,
      this.content,
      required this.createdAt,
      required this.status,
      required this.postModel});
  late final String id;
  late String userId;
  late String postId;
  late String? content;
  late final DateTime createdAt;
  late String status;
  late final PostModel postModel;

  ReportModel.fromJson(Map<String, dynamic> json) {
    id = json['id'];
    userId = json['userId'];
    postId = json['postId'];
    content = json['content'];
    createdAt = DateTime.parse(json['createdAt']);
    status = json['status'];
    postModel = PostModel.fromJson(json["post"]);
  }
}
